pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -811,6 +814,47 @@ func (og *operationGenerator) GenerateMountVolumeFunc(
 	}
 }
 
+func (og *operationGenerator) markDeviceErrorState(volumeToMount VolumeToMount, devicePath, deviceMountPath string, operationState volumetypes.OperationStatus, actualStateOfWorld ActualStateOfWorldMounterUpdater) {
+	switch operationState {
+	case volumetypes.OperationInProgress:
+		// only devices which are not mounted can be marked as uncertain. We do not want to mark a device
+		// which was previously marked as mounted here as uncertain.
+		if actualStateOfWorld.GetDeviceMountState(volumeToMount.VolumeName) == DeviceNotMounted {
+			markDeviceUncertainError := actualStateOfWorld.MarkDeviceAsUncertain(volumeToMount.VolumeName, devicePath, deviceMountPath)
+			if markDeviceUncertainError != nil {
+				klog.Errorf(volumeToMount.GenerateErrorDetailed("MountDevice.MarkDeviceAsUncertain failed", markDeviceUncertainError).Error())
+			}
+		}
+	case volumetypes.OperationFinished:
+		// Similarly only devices which were uncertain can be marked as unmounted
```

Comment:
it might still have issue?
1. mount time out, mark as uncertain
2. second mount returns Finished due to some reason such as secret is deleted
3. mount succeeded?

Reply:
I think we can simply do nothing (do not mark as mounted, nor mark it as unmounted) if volumetype is operationFinished for (re)mount

Reply:
So basically same mount operation which was in-progress in step#1 failed in step#2 and then mount operation initiated in step#1 succeeds.   I agree that this could happen but:

1. for CSI drivers, they can't return a "final" error code for a in-progress operation and other code paths which use external API return `NoChange` status.
2. For secrets etc - I am going to make it to return `NoChange` on `remount` if Secret can not be found, as jan proposed in - https://github.com/kubernetes/kubernetes/pull/82492#discussion_r345795525 but since "uncertain" is never returned for secrets in general, the likelihood of this happening is low and will fixed for good in the follow up of secrets/configmaps PR.

Reply:
Is there any bad consequence of leaving the volume as "uncertain"?

Reply:
I can't think of any bad consequence(atm) except - I have tried to stick to existing data model used by volumemanager. In this model - if a volume is never successfully mounted, it is not recorded in ASOW. With above change you are proposing, known failed mounts will be kept in ASOW and should be unmounted/toredown when pod is deleted.

Reply:
Agree your point. I just feel this way it might be safer and simpler. Otherwise, we have to make sure driver always return correct status just introduced. In previous code, successful mount will mark volume as mounted, failed mount does not unmark anything (of course, there is actually nothing to unmark). 

I think I am ok either way.